### PR TITLE
Add typst language and lsp

### PR DESCRIPTION
**Pull request by AlexanderBrevig**
image

Will add to the wiki when/if merged.

Closes https://github.com/helix-editor/helix/issues/6688

### CODE REVIEW

**Comment by AlexanderBrevig** [comment]
https://typst.app/docs/

File: languages.toml
Comment (AlexanderBrevig) [inline languages.toml:1]: This is formatted by `taplo` using helix default.

Maybe it should get formatted?

Let me know, and I can check it out from git and stop the format from happening.

Comment (AlexanderBrevig) [reply languages.toml:1]: Done

File: languages.toml [hunk 1]
```diff
@@ -2635,3 +2636,18 @@ file-types = ["webc"]
...
+injection-regex = "typst"
+file-types = ["typst", "typ"]
+roots = []
+comment-token = "//"
+language-servers = ["typst-lsp"]
+indent = { tab-width = 4, unit = "    " }
+
+[[grammar]]
+name = "typst"
+# source = { path = "/home/alexander/github.com/tree-sitter-typst" }
```
Comment (the-mikedavis) [inline languages.toml:2652]:
```suggestion
```

Comment (AlexanderBrevig) [reply languages.toml]: Fixed, thanks/sorry

File: runtime/queries/typst/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,9 @@
+([(show_expression) (set_expression)] @constant)
+([(line_comment) (block_comment)] @comment)
+((identifier) @function)
+((string_literal) @string)
+((asssigned_argument) @type)
+(["#"] @punctuation)
+(["+"] @operator)
+(["(" ")" "{" "}" "[" "]"] @punctuation.brackets)
```
Comment (the-mikedavis) [inline runtime/queries/typst/highlights.scm:8]:
```suggestion
(["(" ")" "{" "}" "[" "]"] @punctuation.bracket)
```

**Review by the-mikedavis — DISMISSED** [review]

**Comment by erasin** [comment]
https://github.com/helix-editor/helix/issues/6688

File: languages.toml [hunk 3]
```diff
@@ -2635,3 +2636,17 @@ file-types = ["webc"]
...
+
+[[language]]
+name = "typst"
+scope = "source.typst"
+injection-regex = "typst"
+file-types = ["typst", "typ"]
+roots = []
+comment-token = "//"
+language-servers = ["typst-lsp"]
+indent = { tab-width = 4, unit = "    " }
```
Comment (PerpetualCreativity) [inline languages.toml:2648]: FYI, the typst docs and online editor both use 2 spaces for indentation, not 4. [Here's an example.](https://typst.app/docs/tutorial/advanced-styling/#set-rules)

Comment (AlexanderBrevig) [reply languages.toml]: Thanks! That's true, I've fixed it 👍🏻

**Comment by uben0** [comment]
I am not big on pull request (never done it). But, is it possible to change the grammar to my own on the pull request?
I wrote and maintain a grammar which is complete at https://github.com/uben0/tree-sitter-typst
Or should I create another pull request?

**Comment by fufexan** [comment]
@uben0 either ask @AlexanderBrevig to use your grammar, or open a PR.

You can get this PR's diff [here](https://github.com/helix-editor/helix/pull/7474.diff) and apply it to your helix clone with `git apply -p1 < typst.patch`. Then you can modify the grammar to point to your repo.

**Comment by gabydd** [comment]
Just fyi @uben0 to be able to add it to helix you'll need to check in these files https://github.com/uben0/tree-sitter-typst/blob/master/.gitignore#L5-L8 (all the ones generated by the cli) I have been using your grammar for a couple days and it does seem the most complete so I agree with using it

**Comment by AlexanderBrevig** [comment]
I will update this PR

**Comment by AlexanderBrevig** [comment]
> Just fyi @uben0 to be able to add it to helix you'll need to check in these files https://github.com/uben0/tree-sitter-typst/blob/master/.gitignore#L5-L8 (all the ones generated by the cli) I have been using your grammar for a couple days and it does seem the most complete so I agree with using it

Will you let me know @uben0 when the files are in the repo?
Nice work BTW!

**Comment by uben0** [comment]
Good morning

I added the generated source. If I did something wrong, tell me.

And the highlight and injection queries files are in the queries directory.

Thanks a lot.
------- Original Message -------

**Comment by AlexanderBrevig** [comment]
Updated TS grammar to https://github.com/uben0/tree-sitter-typst

Thank you @uben0!

image

**Comment by uben0** [comment]
Hi,

There was an typo in Cargo.toml, where the grammar had a different name. It was causing a bug when executing cargo test. I fixed it. If you can update the revision number.
------- Original Message -------

File: runtime/queries/typst/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+(math "$" @operator)
+"#" @operator
+"end" @operator
+
+(escape) @constant.character.escape
+["(" ")" "{" "}"] @ponctuation.bracket
+["," ";" ".." ":" "sep"] @ponctuation.delimiter
+"assign" @ponctuation
+(field "." @ponctuation)
```
Comment (the-mikedavis) [inline runtime/queries/typst/highlights.scm:79]:
```suggestion
["(" ")" "{" "}"] @punctuation.bracket
["," ";" ".." ":" "sep"] @punctuation.delimiter
"assign" @punctuation
(field "." @punctuation)
```

**Review by realeinherjar — APPROVED** [review]
LGTM, just get the @the-mikedavis changes in

File: runtime/queries/typst/highlights.scm (hunk 4 above)
Comment (AlexanderBrevig) [reply runtime/queries/typst/highlights.scm]: fixed

**Review by the-mikedavis — APPROVED** [review]

File: runtime/queries/typst/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+(bool) @constant.builtin.boolean
+(builtin) @constant.builtin
+(none) @constant.builtin
+(auto) @constant.builtin
+(ident) @variable
+(call
+  item: (builtin) @function.builtin)
+
+; MARKUP
+(item "item" @operator)
```
Comment (lydiandy) [inline runtime/queries/typst/highlights.scm:56]: @AlexanderBrevig  56,57 reports an error with latest tree-sitter-typst.
```
(item "item" @operator) 
(term ["item" ":"] @operator)
```
here is the error message:
```
". Are your grammars out of sync? Try running 'hx --grammar fetch' and 'hx --grammar build'. This query could not be parsed: QueryError { row: 56, column: 8, offset: 1424, message: "
```

**Review by the-mikedavis — APPROVED** [review]

**Comment by realeinherjar** [comment]
Too bad that this could not be in 23.11 :(

**Comment by lydiandy** [comment]
> Too bad that this could not be in 23.11 :(
I try it again in the latest hx.
it is ok.

**Review by pascalkuthe — APPROVED** [review]

**Comment by cinerea0** [comment]
> Too bad that this could not be in 23.11 :(

Agreed, given the five months between the previous release and the latest one.

**Comment by Idobenhamo** [comment]
@AlexanderBrevig Maybe it is worth it to update the tree sitter to the newest version to support typst 0.9

**Comment by uben0** [comment]
I have been checking the change log for Typst 0.9 and it seams like the only change related to syntax is the addition of builtin `quote` function. I can't find any other change related to syntax. Can anyone double check this?

**Pull request merged**